packages/react-storage/src/components/StorageBrowser/context/controls/Navigate/Navigate.tsx

```diff
@@ -40,13 +41,53 @@ export function navigateReducer(
   state: NavigateState,
   action: NavigateAction
 ): NavigateState {
-  if (action.type === 'SELECT_LOCATION') {
-    return {
-      ...state,
-      location: { ...state.location, current: action.location },
-    };
+  switch (action.type) {
+    case 'SELECT_LOCATION':
+      return {
+        location: {
+          current: action.location,
+          shouldRefresh: false,
+        },
+        history: {
+          list: undefined,
+          shouldRefresh: true,
+        },
+      };
+    case 'DESELECT_LOCATION':
+      return {
+        location: {
+          current: undefined,
+          shouldRefresh: true,
+        },
+        history: {
+          list: undefined,
+          shouldRefresh: false,
+        },
+      };
+    case 'ENTER_FOLDER': {
+      const { name } = action;
+
+      return {
+        ...state,
+        history: {
+          list: [...(state.history.list ?? []), name],
+          shouldRefresh: true,
+        },
+      };
+    }
+    case 'EXIT_FOLDER': {
+      const { index } = action;
+      const updatedHistory = state.history.list?.slice(0, index + 1);
+
+      return {
+        ...state,
+        history: {
+          list: updatedHistory,
+          shouldRefresh: true,
+        },
+      };
+    }
```

Comment:
Adding values to our control state increases likelihood of long term complexity and tech debt due to  exponential increases in possible event variations and usage by consumers, and in this case impacts the `StorageBrowser` interface which needs to be as minimal as possible.

A value conveying intent (e.g. `shouldRefresh`) has to be "cleaned up" (e.g. set to `false`) after any side effects are run to avoid non-deterministic behavior in state consumers.

On selection of a `location`, we rely on React to be React - updating `state.location.current` with a new  reference signals the `useEffect` in `Controller` responsible for fetching `location` data to run with the new values inside `location.state.current`. If `state` updates from a different action, but `state.location` stays untouched, the `useEffect` will ignore the `state` update - no need for a `shouldRefresh` flag.

Think there are probably some simplifications we can do here 😅